neat_ml/opencv/detection.py

```diff
@@ -0,0 +1,210 @@
+    img_path: str,
+    bubble_data: pd.DataFrame,
+    out_dir: Path,
+) -> None:
+    """
+    Create and save a side-by-side figure containing the original
+    image next to the image overlaid with opencv segmentation mask
+
+    Parameters
+    ----------
+    img_path : str
+        File path to the original image.
+    bubble_data : pd.DataFrame
+        DataFrame of bubble metadata as returned by _detect_single_image.
+    out_dir : Path
+        Directory where the debug PNG will be saved.
+    """
+    image_gray = cv2.imread(img_path, cv2.IMREAD_GRAYSCALE)
+
+    if image_gray is None:
+        raise FileNotFoundError(f"Could not read image for debug overlay: {img_path}")
+    
+    image_rgb = skimage.color.gray2rgb(image_gray)
+
+    overlay = image_gray.copy()
+    overlay = cv2.cvtColor(overlay, cv2.COLOR_GRAY2RGB)
+
+    for index, bubble in bubble_data.iterrows():
+        bbox = bubble["bbox"]
+        x_min, y_min, x_max, y_max = map(int, bbox)
+        cv2.rectangle(overlay, (x_min, y_min), (x_max, y_max), (0, 255, 0), 2)
+
+    fig, ax = plt.subplots(1, 2, figsize=(12, 8))
+    ax[0].imshow(image_rgb)
+    ax[0].set_title("Original")
+    ax[0].axis("off")
+    ax[1].imshow(overlay)
+    ax[1].set_title("Detected blobs")
+    ax[1].axis("off")
+
+    png_name = f"{Path(img_path).stem}_debug.png"
+    fig.savefig(out_dir / png_name, dpi=300, bbox_inches="tight")
+    plt.close(fig)
+
+def run_opencv(
+    df: pd.DataFrame,
+    output_dir: str | Path,
+    debug: bool = False,
+) -> pd.DataFrame:
+    """
+    Detect bubbles in every image referenced by df using the
+    OpenCV ``SimpleBlobDetector``.
+
+    Parameters
+    ----------
+    df : pd.DataFrane
+        dataframe containing absolute image filepaths.
+    output_dir : str | Path
+        Path to save the outputs.
+    debug : bool
+        If True save side by side diagnostic images.
+
+    Returns
+    -------
+    df_out : pandas.DataFrame
+        Copy of df enriched with: 
+            num_blobs_opencv number of blobs detected
+            median_radii_opencv median droplet radius
+    """
+    if 'image_filepath' not in df.columns:
+        raise ValueError("DataFrame must contain 'image_filepath' column.")
+    
+    out_dir = Path(output_dir).expanduser().resolve()
+    out_dir.mkdir(parents=True, exist_ok=True)
+
+    memory = Memory(location=str(out_dir / ".joblib_cache"), verbose=0)
+    cached_detect = memory.cache(_detect_single_image)
+
+    df_out = df.copy()
+    df_out[["num_blobs_opencv", "median_radii_opencv"]] = np.nan
+   
+    for idx, row in tqdm(
+        df_out.iterrows(),
+        total=df_out.shape[0],
+        desc="OpenCV SimpleBlobDetector",
+    ):
+        img_path = str(row.image_filepath)
+        image_basename = Path(img_path).stem
+
+        num_blobs, median_r, bubble_data = cached_detect(img_path)
+
+        df_bubbles = pd.DataFrame(bubble_data)
+        df_bubbles.to_pickle(out_dir / f"{image_basename}_bubble_data.pkl")
```

Comment:
Typically one would want to avoid pickle for various reasons, most notably portability between Python versions. `parquet` or even CSV may suffice, depending on data size.